### PR TITLE
feat: Add report to user posting rights

### PR DESCRIPTION
**Pull request by attiks**
Refs: #RW-1058, #RW-1059

### CODE REVIEW

File: html/modules/custom/reliefweb_entities/src/Entity/Report.php [hunk 1]
```diff
@@ -249,6 +249,14 @@ public function preSave(EntityStorageInterface $storage) {
...
     // Prepare notifications.
     $this->preparePublicationNotification();
+
+    // Update the entity status based on the user posting rights.
+    $this->updateModerationStatusFromPostingRights();
+
+    // Update the entity status based on the source(s) moderation status.
+    $this->updateModerationStatusFromSourceStatus();
+
+    parent::preSave($storage);
```
Comment (orakili) [inline html/modules/custom/reliefweb_entities/src/Entity/Report.php:259]: Needs some tests to confirm that moving the preSave at the end of the function doesn't break anything.

File: html/modules/custom/reliefweb_moderation/src/Services/ReportModeration.php [hunk 2]
```diff
@@ -221,30 +224,81 @@ public function getFilterDefaultStatuses() {
...
+          '#value' => $this->t('Reference'),
+        ],
+        'archive' => [
+          '#value' => $this->t('Archive'),
+        ],
+      ];
+    }
+    // Other users can submit for review (or publish directly if trusted).
+    else {
+      $buttons['pending'] = [
```
Comment (orakili) [inline html/modules/custom/reliefweb_moderation/src/Services/ReportModeration.php:282]: Reports do not have a pending status (it's added in the POST API). This needs to be added to the list of allowed statuses. See https://github.com/UN-OCHA/rwint9-site/pull/723/files#diff-8c045d8e06004cb32ea3cdfdde3e25448dd0ce3fc8873f1e0c77237a54c33061R221 and related code.

BUT the needs for this status should be confirmed with the editorial team.

File: html/modules/custom/reliefweb_moderation/src/Services/ReportModeration.php [hunk 3]
```diff
@@ -221,30 +224,81 @@ public function getFilterDefaultStatuses() {
...
+        '#value' => $new ? $this->t('Submit') : $this->t('Submit changes'),
+      ];
+
+      // Add confirmation when attempting to change published document.
+      if ($status === 'published' || $status === 'expired') {
```
Comment (orakili) [inline html/modules/custom/reliefweb_moderation/src/Services/ReportModeration.php:287]: Reports do not have an `expired` status (and do not need one).

File: html/modules/custom/reliefweb_moderation/src/Services/ReportModeration.php [hunk 4]
```diff
@@ -221,30 +224,81 @@ public function getFilterDefaultStatuses() {
...
+        $message = $this->t('Press OK to submit the changes for review by the ReliefWeb editors. The job may be set as pending.');
```
Comment (orakili) [inline html/modules/custom/reliefweb_moderation/src/Services/ReportModeration.php:288]: This is mentioning `job`.

File: html/modules/custom/reliefweb_moderation/src/Services/ReportModeration.php [hunk 5]
```diff
@@ -221,30 +224,81 @@ public function getFilterDefaultStatuses() {
...
+        $buttons['pending']['#attributes']['onclick'] = 'return confirm("' . $message . '")';
+      }
+    }
+
+    // Warning message when saving as a draft.
+    if (isset($buttons['draft'])) {
+      $message = $this->t('You are saving this document as a draft. It will not be visible to visitors. If you wish to proceed with the publication kindly click on @buttons instead.', [
```
Comment (orakili) [inline html/modules/custom/reliefweb_moderation/src/Services/ReportModeration.php:295]: This should not be displayed to editors. This is just going to slow them down.

File: html/modules/custom/reliefweb_user_posts/src/Form/UserPostsPageFilterForm.php [hunk 6]
```diff
@@ -30,13 +30,16 @@ public function buildForm(array $form, FormStateInterface $form_state, Moderatio
     // Link to create a new entity.
     $url_options = ['attributes' => ['target' => '_blank']];
 
-    $links = $this->t('Create a new <a href="@job_url">Job vacancy</a> or a new <a href="@training_url">Training program</a>', [
+    $links = $this->t('Create a new <a href="@job_url">Job vacancy</a>, a new <a href="@training_url">Training program</a> or a new <a href="@report_url">Report</a>', [
```
Comment (orakili) [inline html/modules/custom/reliefweb_user_posts/src/Form/UserPostsPageFilterForm.php:33]: This is some conditional checks. Only people with Contributor role should be able to create reports, not everybody.

File: html/modules/custom/reliefweb_user_posts/src/Form/UserPostsPageFilterForm.php [hunk 7]
```diff
@@ -30,13 +30,16 @@ public function buildForm(array $form, FormStateInterface $form_state, Moderatio
...
       '@job_url' => Url::fromRoute('node.add', [
         'node_type' => 'job',
       ], $url_options)->toString(),
       '@training_url' => Url::fromRoute('node.add', [
         'node_type' => 'training',
       ], $url_options)->toString(),
+      '@report_url' => Url::fromRoute('node.add', [
```
Comment (orakili) [inline html/modules/custom/reliefweb_user_posts/src/Form/UserPostsPageFilterForm.php:40]: Same a above comment: only display for users with appropriate permissions.

**Review by orakili — CHANGES_REQUESTED** [review]
This allows anybody to create reports. The posting rights only impact the status when saving them.

The buttons and statuses need some review and adjustments for the reports.

File: html/modules/custom/reliefweb_moderation/src/Services/ReportModeration.php (hunk 5 above)
Comment (attiks) [reply html/modules/custom/reliefweb_moderation/src/Services/ReportModeration.php:295]: Limited to contributor role.

File: html/modules/custom/reliefweb_moderation/src/Services/ReportModeration.php (hunk 2 above)
Comment (attiks) [reply html/modules/custom/reliefweb_moderation/src/Services/ReportModeration.php:283]: Added (for now)

Comment (attiks) [reply html/modules/custom/reliefweb_moderation/src/Services/ReportModeration.php]: And removed

File: html/modules/custom/reliefweb_entities/src/Entity/Report.php (hunk 1 above)
Comment (attiks) [reply html/modules/custom/reliefweb_entities/src/Entity/Report.php:259]: See https://github.com/UN-OCHA/rwint9-site/pull/905

File: html/modules/custom/reliefweb_moderation/src/Services/ReportModeration.php [hunk 8]
```diff
@@ -221,30 +224,83 @@ public function getFilterDefaultStatuses() {
...
     }
+    // Other users can submit for review (or publish directly if trusted).
+    else {
+      $buttons['to-review'] = [
+        '#value' => $new ? $this->t('Submit') : $this->t('Submit changes'),
+      ];
+
+      // Add confirmation when attempting to change published document.
+      if ($status === 'published') {
+        $message = $this->t('Press OK to submit the changes for review by the ReliefWeb editors. The report may be set as to-review.');
```
Comment (orakili) [inline html/modules/custom/reliefweb_moderation/src/Services/ReportModeration.php:288]:
```suggestion
        $message = $this->t('Press OK to submit the changes for review by the ReliefWeb editors. The report may become unpublished while being reviewed.');
```

File: html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php [hunk 9]
```diff
@@ -271,6 +272,7 @@ protected function initFilterDefinitions(array $filters = []) {
       'values' => [
         'job' => $this->t('Job'),
         'training' => $this->t('Training'),
+        'report' => $this->t('Report'),
```
Comment (orakili) [inline html/modules/custom/reliefweb_user_posts/src/Services/UserPostsService.php:275]: Only show to contributors?

**Review by orakili — CHANGES_REQUESTED** [review]
With the current implementation, a user with the Contributor role can post a report for any source regardless of their posting rights. The report will be saved as `to-review`.

Submission should probably be either not be allowed (form validation failure) or saved as `on-hold` with a flag indicating the user has no posting rights for the selected sources.

The contributor role is missing some permissions. 

For example, it's not possible to upload an image (no permission to access media)

PDF previews are also not being displayed after uploading in the form, probably because of a lack of permission as well (no permission to access to private files).

No permission to view the revisions or the moderation status (at the top of the report) but that may be fine.

File: config/system.action.user_add_role_action.contributor_role.yml [hunk 10]
```diff
@@ -0,0 +1,14 @@
+uuid: 8a1c6d2a-cf03-48d7-9dc8-ba12b7239159
+langcode: en
+status: true
+dependencies:
+  config:
+    - user.role.contributor_role
+  module:
+    - user
+id: user_add_role_action.contributor_role
+label: 'Add the Contributor role role to the selected user(s)'
```
[the remark this answers is not in the document]
Comment (attiks) [reply config/system.action.user_add_role_action.contributor_role.yml]: Renamed

**Comment by attiks** [comment]
See https://github.com/UN-OCHA/rwint9-site/pull/908

**Pull request closed without merging**